llvm/lib/Target/X86/X86RegisterInfo.cpp

```diff
@@ -1082,8 +1088,44 @@ bool X86RegisterInfo::getRegAllocationHints(Register VirtReg,
       VirtReg, Order, Hints, MF, VRM, Matrix);
 
   unsigned ID = RC.getID();
-  if (ID != X86::TILERegClassID)
+
+  if (ID != X86::TILERegClassID) {
+    if (!VRM || DisableRegAllocNDDHints)
+      return BaseImplRetVal;
+
+    // Add any two address hints after any copy hints.
+    SmallSet<unsigned, 4> TwoAddrHints;
+
+    auto tryAddNDDHint = [&](const MachineOperand &MO) -> void {
+      Register Reg = MO.getReg();
+      Register PhysReg =
+          Register::isPhysicalRegister(Reg) ? Reg : Register(VRM->getPhys(Reg));
+      if (PhysReg && !MRI->isReserved(PhysReg) && !is_contained(Hints, PhysReg))
+        TwoAddrHints.insert(PhysReg);
+    };
+
+    for (auto &MO : MRI->reg_nodbg_operands(VirtReg)) {
+      const MachineInstr &MI = *MO.getParent();
+      if (X86::getNonNDVariant(MI.getOpcode())) {
+        unsigned OpIdx = MI.getOperandNo(&MO);
+        if (OpIdx == 0 && MI.getOperand(1).isReg()) {
```

Comment:
Make `MI.getOperand(1).isReg()` as a assertion instead of a check.

Reply:
[18754fd](https://github.com/llvm/llvm-project/pull/98603/commits/18754fdfe4ede2dd6b4722774cb9228b246dd358)